ui/appDisplay.js

```diff
@@ -85,57 +85,80 @@ class HackPopupMenuItem extends PopupMenu.PopupBaseMenuItem {
+var HackPopupMenu = class extends PopupMenu.PopupMenu {
+    constructor(...params) {
+        super(...params);
+        this.box.add_style_class_name('hack-tooltip');
+        this.actor.add_style_class_name('hack-tooltip-arrow');
+    }
+
+    _onKeyPress(actor, event) {
+        // Don't show this popup menu on key press
+        return Clutter.EVENT_PROPAGATE;
+    }
+}
+
+function createInfoPopup(appIcon) {
+    if (appIcon._infoPopup) {
+        return;
+    }
+
+    appIcon._infoPopupId = null;
+    appIcon._infoPopup = new HackPopupMenu(appIcon, 0.5, St.Side.TOP, 0);
+    appIcon._infoMenuItem = new HackPopupMenuItem();
+    appIcon._infoPopup.addMenuItem(appIcon._infoMenuItem);
+
+    appIcon._infoPopup.actor.hide();
+    Main.uiGroup.add_actor(appIcon._infoPopup.actor);
+
+    appIcon._infoPopup.connect('destroy', () => {
+        if (appIcon._infoPopupId) {
+            GLib.source_remove(appIcon._infoPopupId);
+            appIcon._infoPopupId = null;
         }
     });
 
-    this.track_hover = true;
-    return this.connect('notify::hover', () => {
-        if (this.hover) {
+    appIcon.connect('destroy', () => destroyInfoPopup(appIcon));
+
+    appIcon.track_hover = true;
+    appIcon._tooltipHandler = appIcon.connect('notify::hover', () => {
+        if (appIcon.hover) {
             const inApps = Main.overview.viewSelector.getActivePage() === ViewPage.APPS;
             if (!inApps) {
                 return;
             }
 
-            if (this._infoPopupId)
-                GLib.source_remove(this._infoPopupId);
+            if (appIcon._infoPopupId)
+                GLib.source_remove(appIcon._infoPopupId);
 
-            this._infoPopupId = GLib.timeout_add(
+            appIcon._infoPopupId = GLib.timeout_add(
                 GLib.PRIORITY_DEFAULT,
                 300, () => {
-                    this._infoPopupId = null;
-                    if (this._infoPopup)
-                        this._infoPopup.open();
+                    appIcon._infoPopupId = null;
+                    if (appIcon._infoPopup)
+                        appIcon._infoPopup.open();
                 });
         } else {
-            if (this._infoPopupId) {
-                GLib.source_remove(this._infoPopupId);
-                this._infoPopupId = null;
+            if (appIcon._infoPopupId) {
+                GLib.source_remove(appIcon._infoPopupId);
+                appIcon._infoPopupId = null;
             }
-            if (this._infoPopup)
-                this._infoPopup.close();
+            if (appIcon._infoPopup)
+                appIcon._infoPopup.close();
         }
     });
 }
 
-function destroyInfoPopup(handler) {
-    this._infoPopup.destroy();
-    this._infoPopupId = null;
-    this._infoPopup = null;
-    this._infoMenuItem = null;
-    this.disconnect(handler);
+function destroyInfoPopup(appIcon) {
+    if (!appIcon._infoPopup)
+        return;
+
+    appIcon._infoPopup.destroy();
+    appIcon._infoPopupId = null;
```

Comment:
I believe we should remove this timeout source here also otherwise it may try to access `appInfo._infoPopup` that got destroyed above.

Reply:
Good catch. I've removed the timeout in the destroy function if it is not null.